lib/cudadrv/memory.jl

```diff
@@ -485,7 +485,7 @@ function unsafe_copy2d!(dst::Union{Ptr{T},CuPtr{T},CuArrayPtr{T}}, dstTyp::Type{
 
     srcMemoryType, srcHost, srcDevice, srcArray = if srcTyp == HostMemory
         CU_MEMORYTYPE_HOST,
-        src::Ptr,
+        reinterpret(Ptr{T}, src),
```

Comment:
The API accepts multiple pointer types, but not all are viable in all situations. Now you're enabling the possibility of using a CuPtr as HostMemory, which isn't valid.

Reply:
Hmmm, but if we set `default_memory = "host"` in `LocalPreferences.toml`, isn't it the case that the `CuArray` is backed by `HostMemory` but still has a `CuPtr`?

Reply:
`HostMemory` resides on the CPU, so should be accessed via its `Ptr`. Deriving a `CuPtr` calls `cuMemHostGetDevicePointer`, which may be a different pointer than the CPU-accessible one.

I guess it's not ideal that the default `pointer(::CuArray)` method always returns a `CuPtr`, but that design makes sure you can always use those pointers on the GPU (even if that may be slow, as would be the case with the `CuPtr` you get from `HostMemory`).

Reply:
Hmmm, so if the `M` for the `CuArray` is `HostMemory`, we should then use `pointer(arr.mem)` or something similar?

Reply:
Note that for a native pointer, you can pass the memory type to `pointer`:
```julia-repl
julia> pointer(CuArray{Int,1,CUDA.HostMemory}(undef, 1))
CuPtr{Int64}(0x00007eb6cb000200)

julia> pointer(CuArray{Int,1,CUDA.HostMemory}(undef, 1); type=CUDA.HostMemory)
Ptr{Int64} @0x00007eb6cb000400
```

Reply:
OK, should be fixed now